spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/BindingService.java

```diff
@@ -188,4 +274,39 @@ private void validate(Object properties) {
 			throw new IllegalStateException(dataBinder.getBindingResult().toString());
 		}
 	}
+
+	private static class LateBinding<T> implements Binding<T> {
+
+		private volatile Binding<T> delegate;
+
+		private volatile boolean unbound;
+
+		LateBinding() {
+			super();
+		}
+
+		public synchronized void setDelegate(Binding<T> delegate) {
+			if (this.unbound) {
+				delegate.unbind();
+			}
+			else {
+				this.delegate = delegate;
+			}
+		}
+
+		@Override
+		public synchronized void unbind() {
```

Comment:
if some thread calls `unbind` on the `LateBinding` object (since the reference is already returned from consumer binding) and `setDelegate` is called through retries at around same time, isn't there a chance that `unbind` could potentially be called on the same binding delegate object twice?

Reply:
That's why the methods are synchronized...

- if the successful bind wins, we set the field and the unbind will unbind after it gets the monitor.
- if the unbind wins, we set the boolean and when the bind gets the monitor it will unbind instead of setting the field.